x/rollapp/keeper/block_height_to_finalization_queue.go

```diff
@@ -2,62 +2,81 @@ package keeper
 
 import (
 	"fmt"
+	"slices"
 
 	"github.com/cosmos/cosmos-sdk/store/prefix"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/osmosis-labs/osmosis/v15/osmoutils"
+
 	common "github.com/dymensionxyz/dymension/v3/x/common/types"
 	"github.com/dymensionxyz/dymension/v3/x/rollapp/types"
-	"github.com/osmosis-labs/osmosis/v15/osmoutils"
 )
 
 // FinalizeRollappStates is called every block to finalize states when their dispute period over.
-func (k Keeper) FinalizeRollappStates(ctx sdk.Context) error {
+func (k Keeper) FinalizeRollappStates(ctx sdk.Context) {
 	if uint64(ctx.BlockHeight()) < k.DisputePeriodInBlocks(ctx) {
-		return nil
+		return
 	}
 	// check to see if there are pending  states to be finalized
 	finalizationHeight := uint64(ctx.BlockHeight() - int64(k.DisputePeriodInBlocks(ctx)))
 	pendingFinalizationQueue := k.GetAllFinalizationQueueUntilHeightInclusive(ctx, finalizationHeight)
 
-	return osmoutils.ApplyFuncIfNoError(ctx,
-		// we trap at this granularity because we want to avoid iterating inside the
-		// wrapped function (above), and we want the function to be atomic, so
-		// we cannot trap inside finalizePending
-		func(ctx sdk.Context) error {
-			return k.finalizePending(ctx, pendingFinalizationQueue)
-		})
+	k.finalizePending(ctx, pendingFinalizationQueue)
 }
 
-func (k Keeper) finalizePending(ctx sdk.Context, pendingFinalizationQueue []types.BlockHeightToFinalizationQueue) error {
+func (k Keeper) finalizePending(ctx sdk.Context, pendingFinalizationQueue []types.BlockHeightToFinalizationQueue) {
 	// Iterate over all the pending finalization queue
 	for _, blockHeightToFinalizationQueue := range pendingFinalizationQueue {
 		// finalize pending states
+		rollappsToRemoveFromQueue := make([]string, 0, len(blockHeightToFinalizationQueue.FinalizationQueue))
 		for _, stateInfoIndex := range blockHeightToFinalizationQueue.FinalizationQueue {
-			stateInfo := k.MustGetStateInfo(ctx, stateInfoIndex.RollappId, stateInfoIndex.Index)
-			if stateInfo.Status != common.Status_PENDING {
-				panic(fmt.Sprintf("invariant broken: stateInfo is not in pending state: rollapp: %s: status: %s", stateInfoIndex.RollappId, stateInfo.Status))
-			}
-			stateInfo.Finalize()
-			// update the status of the stateInfo
-			k.SetStateInfo(ctx, stateInfo)
-			// update the LatestStateInfoIndex of the rollapp
-			k.SetLatestFinalizedStateIndex(ctx, stateInfoIndex)
-			// call the after-update-state hook
-			keeperHooks := k.GetHooks()
-			err := keeperHooks.AfterStateFinalized(ctx, stateInfoIndex.RollappId, &stateInfo)
+			err := osmoutils.ApplyFuncIfNoError(ctx,
+				func(ctx sdk.Context) error {
+					return k.finalizePendingState(ctx, stateInfoIndex)
+				})
 			if err != nil {
-				return fmt.Errorf("after state finalized: %w", err)
+				k.Logger(ctx).Error("failed to finalize state", "rollapp", stateInfoIndex.RollappId, "index", stateInfoIndex.Index, "error", err.Error())
+				continue
```

Comment:
Can you add a comment, either here, or maybe as the function docstring, which explains the high level philosophy here about how errors are handled? Like the full lifecycle.

For example, what happens if the rollapp doesnt finalize over and over?
Needs some explanation and guidance, otherwise I'm reading the code and it just seems super error prone and I'm not sure what the assumptions and guarantees are

Reply:
Basically, like, even though the code may be correct, it's not obvious that it is. There is some product assumption here